hazelcast/src/main/java/com/hazelcast/dataconnection/databasediscovery/impl/MSSQLDatabaseDiscovery.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.dataconnection.databasediscovery.impl;
+
+import com.hazelcast.dataconnection.DataConnectionResource;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import static com.hazelcast.dataconnection.impl.JdbcDataConnection.OBJECT_TYPE_TABLE;
+
+public class MSSQLDatabaseDiscovery {
+
+    List<String> systemSchemaList = List.of("sys", "INFORMATION_SCHEMA");
+    List<String> systemTableList = List.of("spt_", "MSreplication_options");
+
+    public List<DataConnectionResource> listResources(Connection connection) throws SQLException {
+        try (ResultSet tables = connection.getMetaData().getTables(
+                     connection.getCatalog(),
+                     null,
+                     null,
+                     new String[]{"TABLE", "VIEW"})) {
+            List<DataConnectionResource> result = new ArrayList<>();
+            while (tables.next()) {
+                // Format DataConnectionResource name as catalog + schema+ + table_name
+                String[] name = Stream.of(
+                                tables.getString("TABLE_CAT"),
+                                tables.getString("TABLE_SCHEM"),
+                                tables.getString("TABLE_NAME")
+                        )
+                        .filter(Objects::nonNull)
+                        .toArray(String[]::new);
+
+                if (isSystemSchema(name[1]) || isSystemTable(name[2])) {
```

Comment:
you don't know if there will be 3 elements. If you know from some invariant about MSSQL driver, please use `checkState()` to explicitly state the invariant - otherwise it's not clear if it's by design or by mistake

Reply:
rev : 70a8a5eef0b40f1bfb29d76cc12e125764b8bffc
This adds FullyQualifiedTableName class which handles 3 elements of fully qualified table name